src/Runtime/Distributions/Automata/Automaton.Builder.cs

```diff
@@ -0,0 +1,810 @@
+                        }
+
+                        endState.SetEndWeight(Weight.Product(endState.EndWeight, secondStartState.EndWeight));
+                    }
+
+                    this.RemoveState(secondStartState.Index);
+                }
+                else
+                {
+                    // Just connect all end states with start state of appended automaton
+                    for (var i = 0; i < oldStateCount; i++)
+                    {
+                        var state = this[i];
+                        if (state.CanEnd)
+                        {
+                            state.AddEpsilonTransition(state.EndWeight, secondStartState.Index, group);
+                            state.SetEndWeight(Weight.Zero);
+                        }
+                    }
+                }
+
+                bool AllEndStatesHaveNoTransitions()
+                {
+                    for (var i = 0; i < oldStateCount; ++i)
+                    {
+                        var state = this.states[i];
+                        if (state.CanEnd && state.FirstTransition != -1)
+                        {
+                            return false;
+                        }
+                    }
+
+                    return true;
+                }
+            }
+
+            #endregion
+
+            #region Result getters
+
+            /// <summary>
+            /// Builds new automaton object.
+            /// </summary>
+            public TThis GetAutomaton() => new TThis() { Data = this.GetData() };
+
+            /// <summary>
+            /// Stores built automaton in pre-allocated <see cref="Automaton{TSequence,TElement,TElementDistribution,TSequenceManipulator,TThis}"/> object.
+            /// </summary>
+            public DataContainer GetData()
+            {
+                if (this.StartStateIndex < 0 || this.StartStateIndex >= this.states.Count)
+                {
+                    throw new InvalidOperationException(
+                        $"Built automaton must have a valid start state. StartStateIndex = {this.StartStateIndex}, states.Count = {this.states.Count}");
+                }
+
+                var hasEpsilonTransitions = false;
+                var resultStates = new StateData[this.states.Count];
+                var resultTransitions = new Transition[this.transitions.Count - this.numRemovedTransitions];
+                var nextResultTransitionIndex = 0;
+
+                for (var i = 0; i < resultStates.Length; ++i)
+                {
+                    var state = this.states[i];
+                    var transitionIndex = state.FirstTransition;
+                    state.FirstTransition = nextResultTransitionIndex;
+                    while (transitionIndex != -1)
+                    {
+                        var node = this.transitions[transitionIndex];
+                        var transition = node.Transition;
+                        Debug.Assert(
```

Comment:
Is there a reason to keep this Debug.Assert in?

Reply:
I don't see why it shouldn't be kept. May be, the same check could be added to `AddTransition()`, I'll take a look.

Reply:
Adding the same check to `AddTransition()` is a bit problematic - in a few places (like FromAutomaton) we actually insert transitions which point to states which weren't inserted yet. It probably is not worth it to restructure code. Because assert in `GetData()` will catch this kind of errors.

I don't see why we shouldn't keep it.